core/src/main/java/io/grpc/Metadata.java

```diff
@@ -362,17 +364,30 @@ private String toStringInternal() {
    *
    * <p>Only the following ASCII characters are allowed in the names of keys:
    * <ul>
-   *   <li>letters, i.e., {@code a-z} and {@code A-Z}</li>
-   *   <li>digits, i.e., {@code 0-9}</li>
-   *   <li>dash, i.e., {@code -}</li>
-   *   <li>underscore, i.e., {@code _}</li>
+   *   <li>letters: {@code a-z}</li>
```

Comment:
Lower-casing is an internal detail; we don't have to do that. We could easily use a TreeMap with String.CASE_INSENSITIVE_ORDER.

The extra memory overhead of maintaining the two strings is negligible. It's not like Key has a ton of fields either, or that we have to do any extra computation to produce that string at any point in the future. So while it could be nice not to keep it, I'm not sure what the concrete concerns are.